src/python/pants/backend/makeself/makeself.py

```diff
@@ -0,0 +1,322 @@
+import os
+from dataclasses import dataclass
+from typing import Optional
+
+from pants.core.util_rules import external_tool
+from pants.core.util_rules.external_tool import (
+    DownloadedExternalTool,
+    ExternalToolRequest,
+    TemplatedExternalTool,
+)
+from pants.core.util_rules.system_binaries import (
+    AwkBinary,
+    Base64Binary,
+    BasenameBinary,
+    BashBinary,
+    BinaryShims,
+    BinaryShimsRequest,
+    Bzip2Binary,
+    CatBinary,
+    ChmodBinary,
+    CksumBinary,
+    CutBinary,
+    DateBinary,
+    DdBinary,
+    DfBinary,
+    DirnameBinary,
+    DuBinary,
+    ExprBinary,
+    FindBinary,
+    GpgBinary,
+    GzipBinary,
+    HeadBinary,
+    IdBinary,
+    Md5sumBinary,
+    MkdirBinary,
+    PwdBinary,
+    RmBinary,
+    SedBinary,
+    ShBinary,
+    SortBinary,
+    TailBinary,
+    TarBinary,
+    TestBinary,
+    TrBinary,
+    WcBinary,
+    XargsBinary,
+    XzBinary,
+    ZstdBinary,
+)
+from pants.engine.fs import Digest, RemovePrefix
+from pants.engine.platform import Platform
+from pants.engine.process import Process, ProcessCacheScope, ProcessResult
+from pants.engine.rules import Get, collect_rules, rule
+from pants.util.logging import LogLevel
+
+logger = logging.getLogger(__name__)
+
+
+class MakeselfSubsystem(TemplatedExternalTool):
+    options_scope = "makeself"
+    help = "A tool to generate a self-extractable compressed tar archives."
+
+    default_version = "2.5.0"
+    default_known_versions = [
+        "2.5.0|macos_arm64 |4d2fa9d898be22c63bb3c6bb7cc3dc97237700dea6d6ad898dcbec0289df0bc4|45867",
+        "2.5.0|macos_x86_64|4d2fa9d898be22c63bb3c6bb7cc3dc97237700dea6d6ad898dcbec0289df0bc4|45867",
+        "2.5.0|linux_arm64 |4d2fa9d898be22c63bb3c6bb7cc3dc97237700dea6d6ad898dcbec0289df0bc4|45867",
+        "2.5.0|linux_x86_64|4d2fa9d898be22c63bb3c6bb7cc3dc97237700dea6d6ad898dcbec0289df0bc4|45867",
+    ]
+
+    default_url_template = "https://github.com/megastep/makeself/releases/download/release-{version}/makeself-{version}.run"
+
+
+@dataclass(frozen=True)
+class RunMakeselfArchive:
+    exe: str
+    input_digest: Digest
+    description: str
+    level: LogLevel = LogLevel.INFO
+    output_directory: Optional[str] = None
+
+
+@rule(desc="Run makeself archive", level=LogLevel.DEBUG)
+async def run_makeself_archive(
+    request: RunMakeselfArchive,
+    awk: AwkBinary,
+    base64: Base64Binary,
+    basename: BasenameBinary,
+    bash: BashBinary,
+    bzip2: Bzip2Binary,
+    cat: CatBinary,
+    cut: CutBinary,
+    dd: DdBinary,
+    df: DfBinary,
+    dirname: DirnameBinary,
+    expr: ExprBinary,
+    find: FindBinary,
+    gpg: GpgBinary,
+    gzip: GzipBinary,
+    head: HeadBinary,
+    id: IdBinary,
+    md5sum: Md5sumBinary,
+    mkdir: MkdirBinary,
+    pwd: PwdBinary,
+    rm: RmBinary,
+    sed: SedBinary,
+    shasum: DateBinary,
+    tail: TailBinary,
+    tar: TarBinary,
+    test: TestBinary,
+    wc: WcBinary,
+    xz: XzBinary,
+    zstd: ZstdBinary,
```

Comment:
Okay, I'm happy enough to leave it for a future PR, but it'd be good to set a direction before you start working on it. Given it sounds like you have an idea in mind, could you sketch out the `BUILD` syntax you're thinking of?

Reply:
Ah, okay. Thanks for explaining.

Our experience from #19643 is strict validation of input parameters to tools that pants is managing can be annoying. For example, do a manual upgrade of the tool using `pants.toml` to have some new functionality (in that case, a new version of pip; in this case, maybe some new compression or new hashing algorithm), and then not be able use it via pants, because pants is validating the arg too strictly.

So, a version that's more resilient to improvements to `makeself` might allow a bit more customisation without needing new pants' versions. Some possible inspiration would be `shell_command` (with its `tools` argument) and `adhoc_tool` (with its `executable_dependencies` & `system_binary`).

Thoughts?